### PR TITLE
Optimization: Escape from useCallback hell

### PR DESCRIPTION
**Pull request by omgovich**
**Discussion/Proposal**

The specs are failing now because they are created for the current realization. But actually, I'm not sure that the current realization is the best we can do.

Since your library returns **callbacks only** it doesn't make sense to update/lose a link to `debouncedState` at all.

So I created a version where debouncedState is always memorized. It also makes code way simpler and saves us 57 bytes.

Thought?

### CODE REVIEW

**Comment by xnimorz** [comment]
Hi @omgovich 

I'm not sure about this technic. 

The biggest problem I see, that users are about to lose control of refs. I'll show the example (not the best and the most correct one, but it shows the problem):
```javascript
import React, { useState, useCallback } from 'react';
import useDebouncedCallback from './useDebouncedCallback';

function SearchForm({ makeSearchRequest, ...anyOtherProps }) {
  const [searchValue, setSearchValue] = useState();
  const denounced = useDebouncedCallback(
    useCallback((e) => {
      makeSearchRequest(e.target.value);
    }, [makeSearchRequest]),
    300
  );
  const searchInputHandler = useCallback((e) => {
    setSearchValue(e);
    denounced.callback(e);
  }, []);
  const submit = useCallback(() => {
    debouned.flush();
  }, [])
  return (
    <Form onSubmit={submit}>
      <Input onChange={searchInputHandler}></input>
      <Button  />
    </Form>
  );
}
```

Here we receive `React Hook useCallback has a missing dependency: 'denounced'. Either include it or remove the dependency array` from `react-hooks/exhaustive-deps`

We are able to fix it properly:
```javascript
import React, { useState, useCallback } from 'react';
import useDebouncedCallback from './useDebouncedCallback';

function SearchForm({ makeSearchRequest, ...anyOtherProps }) {
  const [searchValue, setSearchValue] = useState();
  const denounced = useDebouncedCallback(
    useCallback((e) => {
      makeSearchRequest(e.target.value);
    }, [makeSearchRequest]),
    300
  );
  const searchInputHandler = useCallback((e) => {
    setSearchValue(e);
    denounced.callback(e);
  }, [debounced]);
  const submit = useCallback(() => {
    debouned.flush();
  }, [debounced])
  return (
    <Form onSubmit={submit}>
      <Input onChange={searchInputHandler}></input>
      <Button  />
    </Form>
  );
}
```
But users could destructure our object:
```javascript
import React, { useState, useCallback } from 'react';
import useDebouncedCallback from './useDebouncedCallback';

function SearchForm({ makeSearchRequest, ...anyOtherProps }) {
  const [searchValue, setSearchValue] = useState();
  const {callback, flush} = useDebouncedCallback(
    useCallback((e) => {
      makeSearchRequest(e.target.value);
    }, [makeSearchRequest]),
    300
  );
  const searchInputHandler = useCallback((e) => {
    setSearchValue(e);
    callback(e);
  }, [callback]);
  const submit = useCallback(() => {
    flush();
  }, [flush])
  return (
    <Form onSubmit={submit}>
      <Input onChange={searchInputHandler}></input>
      <Button  />
    </Form>
  );
}
```

We don't have a proper hook to check it in the users' code. It leads the code to the runtime deoptimization: `Input` and `Form` components will be updated each time when `MyComponent` is updated despite `Input` and `Form` could be pure ones.

Solution with `useCallbacks` prevents this. 

Summing up: your idea is a good one, but the problem, that we can't check if users use restructuring or not. It's better to behave as users are expected than save 57 bytes from my point of view

**Comment by omgovich** [comment]
I guess we could solve callbacks memoizing problems.
This change not only about saving bytes, It code simplification and performance as well.

**Comment by omgovich** [comment]
Actually, I tried — it's just PITA to do that with TS, but it's possible of course.

I think that denouncedState shouldn't lose its link and must not rerender components. It runs unnecessary rerenders and makes performance worse.

**Comment by xnimorz** [comment]
I agree, that it could lead to excess rerenderes at the moment, especially when users don't use `useCallback` to cache the function. 

But, with my example the solution based on `useRef`s leads to more rerenders once it used incorrectly.

We can wrap `debounced.callback` in useRef, so that the user will get the same ref each time.
Which resolves described issue.

**Comment by omgovich** [comment]
That's a trick that we use in our picker component.
https://github.com/omgovich/react-colorful/blob/master/src/hooks/useEventCallback.ts

It could help us here, but it's pretty hard to use it with all of your callbacks and TS.

**Comment by xnimorz** [comment]
I'll take a look at the problem a bit later. 
I think we are able not to make useEventCallback, the easier solution could be enough.

Suggest, it could be made with a new major version and promises https://github.com/xnimorz/use-debounce/pull/83

**Comment by tryggvigy** [comment]
Maybe this external api structure would help with this https://github.com/xnimorz/use-debounce/issues/61#issuecomment-688717471? Since the returned thing is a single function (the debounced callback) which can be memoized

```ts
// useDebouncedCallback.tsx

const cancel = useCallback(() => {...});
const flush = useCallback(() => {...});
const pending = useCallback(() => {...});
const debounced = useCallback(() => {...});

debounced.cancel = cancel;
debounced.flush = flush;
debounced.pending = pending;
return debounced;


// MyComponent.tsx

const debounced = useDebouncedCallback(() => {...}, 500);

useEffect(() => {
  if (shouldFlush) {
    debounced.flush();
  }
}, [shouldFlush])

useEffect(debounced.cancel, [])

return <button onClick={debounced}>click me</button>
```

**Comment by xnimorz** [comment]
Yes, It could be. I'll check it, and add some changes to the existing code later

**Comment by tryggvigy** [comment]
Nice, we've been using this pattern in the fork I created that we are using at Spotify and it's quite nice. Since it's not so common to have to use cancel, flush or pending at all it's nice that the common case can be the callback itselt. There is no need to destructure it from the hook and it's memoized. e.g.

```ts
const debounced = useDebouncedCallback(() => {...}, 500);
debounced();

// vs.

const { callback } = useDebouncedCallback(() => {...}, 500);
callback();
```

**Comment by xnimorz** [comment]
I'm closing this PR in favour of https://github.com/xnimorz/use-debounce/pull/68, which also gets rid of useCallbacks
